packages/client/src/execution/vmexecution.ts

```diff
@@ -133,21 +124,93 @@ export class VMExecution extends Execution {
     }
   }
 
-  /**
-   * Run a function after acquiring a lock. It is implied that we have already
-   * initialized the module (or we are calling this from the init function, like
-   * `_setCanonicalGenesisBlock`)
-   * @param action - function to run after acquiring a lock
-   * @hidden
-   */
-  private async runWithLock<T>(action: () => Promise<T>): Promise<T> {
-    try {
-      await this._lock.acquire()
-      const value = await action()
-      return value
-    } finally {
-      this._lock.release()
+  setupMerkleVm() {
+    if (this.merkleVm !== null) {
+      return
+    }
+    const trie = new Trie({
+      db: new LevelDB(this.stateDB),
+      useKeyHashing: true,
+      cacheSize: this.config.trieCache,
+      valueEncoding: this.config.useStringValueTrieDB ? ValueEncoding.String : ValueEncoding.Bytes,
+    })
+
+    this.config.logger.info(`Setting up merkleVm`)
+    this.config.logger.info(`Initializing account cache size=${this.config.accountCache}`)
+    this.config.logger.info(`Initializing storage cache size=${this.config.storageCache}`)
+    this.config.logger.info(`Initializing code cache size=${this.config.codeCache}`)
+    this.config.logger.info(`Initializing trie cache size=${this.config.trieCache}`)
+
+    const stateManager = new DefaultStateManager({
+      trie,
+      prefixStorageTrieKeys: this.config.prefixStorageTrieKeys,
+      accountCacheOpts: {
+        deactivate: false,
+        type: CacheType.LRU,
+        size: this.config.accountCache,
+      },
+      storageCacheOpts: {
+        deactivate: false,
+        type: CacheType.LRU,
+        size: this.config.storageCache,
+      },
+      codeCacheOpts: {
+        deactivate: false,
+        type: CacheType.LRU,
+        size: this.config.codeCache,
+      },
+    })
+    this.merkleVm = new (VM as any)({
+      common: this.config.execCommon,
+      blockchain: this.chain.blockchain,
+      stateManager,
+      profilerOpts: this.config.vmProfilerOpts,
+    })
+  }
+
+  setupVerkleVm() {
+    if (this.verkleVm !== null) {
+      return
+    }
+
+    this.config.logger.info(`Setting up verkleVm`)
+    const stateManager = new StatelessVerkleStateManager()
+    this.verkleVm = new (VM as any)({
+      common: this.config.execCommon,
+      blockchain: this.chain.blockchain,
+      stateManager,
+      profilerOpts: this.config.vmProfilerOpts,
+    })
```

Comment:
Hmm, now that I am looking at this merkle vs verkle VM thing here again: both VMs are initialized in exactly the same way. I think we can remove this distinction completely and just go (back) to `this.vm` and only setup the respective different state manager.

Can you refactor a bit here?

Reply:
Update: ok ok, I see this is for this transition logic. So I see where this is coming from, still can't oversee if this is the optimal/cleanest solution though. Maybe you can also have another look at this code part with the comments here in mind and see if you still think this is logically best structured and either keep or update.